harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -210,6 +348,95 @@ def _set_data_loaders(self) -> None:
                     )
                 self.validation_loader = validation_data
 
+    def _step_batch(self):
+        self.batches_trained += 1
+
+        # True epoch-based training is not supported. Epoch start/end is calculated with batch.
+        if self.context.is_epoch_start():
+            self.on_epoch_start(self.get_epoch_idx(self.batches_trained))
+
+        if self.context.is_epoch_end():
+            self.on_epoch_end(self.get_epoch_idx(self.batches_trained))
+
+        if issubclass(type(self.max_length), Batch):
+            self._train_step()
+
+    def _step_epoch(self):
+        self.epochs_trained += 1
+        if issubclass(type(self.max_length), Epoch):
+            self._train_step()
+
+    def _train_step(self):
+        self.steps_completed += 1
+        if self.steps_completed % self.validation_period == 0:
+            self.report_training_progress(self.searcher_op, self.training_metrics)
+            val_metrics = self._validate()
+            self.report_validation_metrics(self.steps_completed, val_metrics)
+            self.stop_requested()
+            self.check_for_preemption()
+        if self.steps_completed % self.checkpoint_period == 0:
+            self.report_training_progress(self.searcher_op, self.training_metrics)
+            self.checkpoint()
+            self.check_for_preemption()
+
+    def stop_requested(self):
+        if self.context.get_stop_requested():
+            raise ShouldExit()
+
+    def _steps_remaining(self):
+        return self.max_length.value - self.steps_completed
+
+    def _train(self, train_unit, min_checkpoint_period, min_validation_period):
+        self.max_length = train_unit
+        self.validation_period = min_validation_period
+        self.checkpoint_period = min_checkpoint_period
+        while self._steps_remaining() > 0:
+            for batch_idx, batch in enumerate(self.training_loader):
```

Comment:
I think we need to highlight this as a breaking change, because we've previously always told users to [always repeat their training dataset](https://docs.determined.ai/latest/training/apis-howto/api-pytorch-ug.html#customize-a-reproducible-dataset).

Reply:
also, in `def run()` we are setting up `self.training_iterator`, we'll need to rectify that code with this code somehow.

Reply:
also, switching to this strategy is going to require hvd.join() and something specific to torch.distributed.

Reply:
we shouldn't need to support hvd.join or torch.join until we support true-epoch based training. currently the logic for stepping optimizers/lr schedulers conforms with previous epoch end calculations.

Reply:
I don't think that's true.  I added this snippet of code fairly recently:

```        # All workers use the chief's definition of epoch lengths, which is based on the training
        # loader's len. If this len does not exist, epoch durations cannot be deduced, and they
        # default to max int.
        try:
            epoch_len = len(self.training_loader)
        except TypeError:
            epoch_len = sys.maxsize
        self.context._epoch_len = self.context.distributed.broadcast(epoch_len)
```

So on tip-of-master, epoch_end calls are synchronized to what the master thinks is the end-of-epoch.

This is actually a big deal because stepping LR schedulers out-of-sync will desynchronize the replicas.